noir-projects/noir-protocol-circuits/crates/types/src/meta/mod.nr

```diff
@@ -71,7 +85,30 @@ pub comptime fn generate_deserialize_from_fields<N>(
         // The currently processed field should be replaced so we do so
         result = ro_replace_with;
     } else {
-        if typ.is_field() | typ.as_integer().is_some() | typ.is_bool() {
+        // If the type implements `Packable`, its length will be assigned to the `maybe_packed_len_typ` variable.
+        let maybe_packed_len_typ = std::meta::typ::fresh_type_variable();
+        let packable_constraint = quote { Packable<$maybe_packed_len_typ> }.as_trait_constraint();
+
+        if (should_unpack & typ.implements(packable_constraint)) {
+            // Unpacking is enabled and the given type implements the `Packable` trait so we call the `unpack()`
+            // method, add the resulting field array to `aux_vars` and each field to `fields`.
+            let packed_len = maybe_packed_len_typ.as_constant().unwrap();
+
+            // We copy the packed fields into a new array and pass that to the unpack function in a quote
+            let mut packed_fields_quotes = &[];
+            for i in 0..packed_len {
+                let index_in_field_array = i + num_already_consumed;
+                packed_fields_quotes = packed_fields_quotes.push_back(
+                    quote { $field_array_name[$index_in_field_array] },
+                );
+            }
+            let packed_fields = packed_fields_quotes.join(quote {,});
+
+            // No we call unpack on the type
+            result = quote { aztec::protocol_types::traits::Packable::unpack([ $packed_fields ]) };
+
+            consumed_counter = packed_len;
```

Comment:
I'm unsure about this. What if I have `struct Foo { a: u32, b: u32 }`? Even if `u32` implemented packable then this would not do the correct thing (i.e. packing the u32 together). So what's the point of having this if I still need to do it manually?

Reply:
>So what's the point of having this if I still need to do it manually?
Because it works well for the typical use case which is a note containing a U128.

If devs want to do some complex packing I would expect them to do it manually.

But if you want to do a more advanced packing that is measuring size of the types then I would do that in a follow-up PR. But not sure if it's good time to do it as it feels like something we could leave up to devs to experiment first with (will there ever be a note containing multiple u32s)?

Reply:
Because if some of the struct members implements "real packing" (like e.g. U128) then we achieve a smaller result because packing of that struct gets used.

E.g. packing UintNote which is defined as:

```
#[partial_note(quote {value})]
#[derive(Serialize)]
pub struct UintNote {
    // The amount of tokens in the note
    value: U128,
    owner: AztecAddress,
    // Randomness of the note to hide its contents
    randomness: Field,
}
```

resulted in the note content occupying 3 fiedls instead of 4 (because the value got actually packed).

Reply:
Yes, I see. So you'd use the derived version if all members either a) occupy a field or b) are packable. If this is not true (e.g. you have two `u128` members) then you'd roll your own.

Reply:
Exactly. And given that a more sophisticated packing can be introduced just with an update of `Aztec.nr` and that we currently don't have any note type which would leverage it I think it makes sense to not deal with that in the foreseeable future.